### DISCUSSION TITLE
Switching image output from PNG to JPG?

### DISCUSSION THREAD

**Discussion by rkhamilton** [rkhamilton/vqgan-clip-generator#45] [Ideas]
How would you feel about me changing the output from png to jpg? 
Benefits of switching to JPG:
* I could use exif / IPTC metadata fields to store titles and comments that would show up in Windows Explorer / Lightroom / standard photo apps.
* File size savings are significant when extracting stills from video. My 5 second, 24MB test video extracts to 1.12GB. That seems unscalable for styling bigger videos. If I instead extract to JPG, it's only 17MB, 1000-fold smaller.
* Potential big speed increase for metadata copying. The operation to read PNG data chunks and copy them to the upscaled image is incredibly slow. It's as slow as running Real-ESRGAN to upscale the images, or slower. 

Cons:
* Potential reduced quality in output when postprocessing? While PNG is lossless, I don't think we'd lose anything practically because these generated images aren't a pixel-perfect recreation of anything. I don't think the GAN is creating fine gradations of color or fine detail that could be lost. It's much more an impression of the subject. So I'm thinking the benefits of JPG outweigh any loss from switching from PNG.

If I change the output to JPG I wouldn't want to make it an option to keep PNG because I wouldn't want to maintain a big duplicate set of unit tests.

Any concerns?

**Comment by gateway** [comment]
Hmm.. so thats a super tough one especially with my background.  I always prefer to get the highest quality image so when my final output is done its pristine when it comes to video production, or printing my images on large metal.

I understand the reason for this but for people doing big video projects usually have  a decent amount of HD space for anything they are working on.

On keeping the idea of having two formats I understand the complexity that comes with supporting that, esp sing a lot of the code seems to revolve around the naming of the images with a .png extension.  With my web build I convert png's to jpgs due to just some of my png's are 30-40 megs each and just not a viable solution for the web (if you wanted to post on the web).

Ideally decoupling the extension and adding in some if check statements is it a lot of work? I mean if your already going to go though and change everything to jpg it might just be smart to take the extra time to offer both? 

you could then add a extension config option.. and based upon that some if checks to save as the png or jpg.. I know its not as simple as that since scanning the code and docs .png is everywhere.. but you will have to change that all to jpg.. 

my 2 cents but im not the one maintaining the code.

**Reply by rkhamilton** [reply]
The burden for maintaining it as an option isn't about coding it the first time, it's keeping both options tested and working in the future as more stuff changes around it.

If you look in the tests folder, I've got unit tests for as many of the [generation configurations as I reasonably can](https://github.com/rkhamilton/vqgan-clip-generator/blob/main/tests/test_generate.py). It already takes so long to run the full set of tests that I basically run it overnight, which means I don't test often enough. Bugs have slipped through when my unit test would have caught it failing and I just didn't run it because it took too long. And if I were to maintain two output options, I'd have to essentially double the number of tests, the PNG path and the JPG path, to be sure the full toolchain works for both.

People always want software to offer more options, but every option increases the test burden, and adds more code paths that can interact and break in unexpected ways. It's never about the problem of doing it once, it's about keeping it working and not having the dev drown in maintenance (which is not the fun part of coding a hobby project).

Soon-ish I'll start a branch to implement the jpg path and you'll be able to install that dev branch and let me know if you see your workflow falling apart. And it's always possible to install a specific version later, so you could
```sh
pip install git+https://github.com/rkhamilton/vqgan-clip-generator.git@944152e 
```
to get the specific version of the code from today. The stuff after the @ is a commit number. You can see these in the [commits page](https://github.com/rkhamilton/vqgan-clip-generator/commits/main) or the [releases page](https://github.com/rkhamilton/vqgan-clip-generator/releases).

**Reply by sethburnette** [reply]
My kneejerk reaction is to never make any initial output lossy. That can happen later when encoding to video or optimizing for web.

**Reply by rkhamilton** [reply]
I question the assumption that “lossless” output is even meaningful in this context .
* Extracting video frames from a video that is encoded with h.264 is not recovering the original source footage quality. I also struggle to imagine how using that extracted footage as an initial image to generate an image followed by 100 training iterations with VQGAN+CLIP can make any meaningful difference if the source is PNG vs JPG.
* Generating a PNG with generate.image isn’t a lossless representation of a true image encoded in the neural network.

I completely understand the desire for him bit depth, high bit rate source material. I shoot video in log, and photos in raw for that reason, and edit the color of everything that I shoot. You avoid color banding when there are fine color gradations. You get enormously more latitude for adjustments when color grading video or correcting images. I get it.

I question where any of that applies to purely synthetic media. Are there “fine color gradations” in the latent vector neural network representation of the image that would be lost in a high-quality-setting jpg? I don’t think any of us know. And if there are such gradations, I question whether they are lost in a practical sense going to jpg. There are many professional photographers who shoot and edit jpg (at high quality settings) rather than raw, and they don’t lose business because of quality issues.

I’m approaching this from the perspective of an engineer, and I would like to make this change based on data. When I initially implement jpg I’ll start with single image generation and keep png in parallel so we can see them side-by-side. I’ll ask for your help evaluating where there is any practical difference between the two that impacts your workflow. If it turns out to be important I’ll maintain both formats as options.

**Reply by sethburnette** [reply]
I see where you are coming from, for sure. I said "kneejerk" because my first unthinking response when I saw this thread was that I'd hate to generate an otherwise beautiful creation that ends up with baked-in color banding or compression artifacts. Probably not something to worry about though, so go for it!

**Comment by rkhamilton** [comment]
Good news, I pushed [version 2.1](48) with support for jpg, and png, both with metadata.  I was able to better leverage some torch / pillow tools for file formats, and you should be able to export your images in *[any file format supported by pillow](https://pillow.readthedocs.io/en/stable/handbook/image-file-formats.html)*. I have to create custom code to generate metadata to embed in each file format, and I've only done that for png and jpg. If you save in something else like TIFF there won't be any metadata to describe the conditions, but it should generate a valid file for you.

**Reply by gateway** [reply]
Nice job! I appreciate all the work your putting into this project.  I havn't had time to play with much lately but plan on firing up some stuff soon and will have some silly questions probably.. but thank you for doing this..

**Reply by sethburnette** [reply]
The JPG route is so much faster! I take back my initial doubt.

**Reply by gateway** [reply]
:) I'll use jpg for tests but png for my main.. My thread ripper with 2 titans usually doesn't care what I throw at it :)
